Experimental refactor in legion_experiment crate

TomGillen opened the pull request:
This started as a simple refactor of legion into smaller sub-modules, as some of the module files were well over 1k LoC and were getting unmaintainable. However, while moving things around, I found some soundness issues and saw some opportunity to simplify things. Those in turn caused more changes downstream, and the simplifications made other improvements that I had been planning fairly trivial to implement. This ended up being a _much_ larger and more impactful refactor of the entire legion_core crate than I had originally planned. This work is not yet complete, but it is wide reaching enough that I want to give everyone the opportunity to see what I've been doing and collect some feedback.

The general changes are:
 * Much shorter module files, each with a more clearly defined scope.
 * Simpler code in general. I don't have a line count yet, but I expect this to have shaved a few hundred lines off the library overall.
 * Removed many uses of `unsafe`.

Changes to `World` (which now more closely resembles a std collection):
 * Added `len` to retrieve entity count.
 * Renamed `insert` to `extend`.
 * Added `push` as sugar for inserting a single entity.
 * Renamed `delete` to `remove`.
 * Renamed `is_alive` to `contains`.
 * Added `entity(Entity) -> Entry` and `entity_mut(Entity) -> EntryMut` functions. Entries contain the APIs for interacting with a single entity. e.g. `get_component`, `add_component` and `layout` (which in turn offers `has_component` etc).
 * `iter` and `iter_mut` for iterating through all entity entries.
 * You can now insert with a tuple of component vecs, in addition to an iterator of component tuples. This is about 50% faster.

Changes to storage:
 * Much clearer distinction between structural changes to storage (e.g. adding entities) and the interior mutability involved in accessing components.
 * Chunk metadata has been pulled out into a `LayoutIndex`.
 * The layout index is optimised for search and provides a `.search(EntityFilter) -> impl Iterator<Item = ArchetypeIndex>` function.
 * Renamed "archetype" to "layout".
 * Renamed "chunk set" to "archetype".
 * Storage now derefs into `[Archetype]` and can be indexed by `ArchetypeIndex`.
 * Chunks no longer free their internal memory when empty. Defrag now deletes empty chunks.
 * Event subscribers can ask to recieve initial creation/insertion events for pre-existing archetypes and entities when they subscribe.

Changes to filters:
 * No longer implemented as iterators. This was a pain without GATs (it is the canonical streaming iterator problem).
 * Filters now only provide their matching logic.
 * Distinction made between stateless layout/archetype filters and stateful chunk filters.
 * Filters can now _cache their matches_ and incrementally update queries. System queries should do this by default in the future.

It is likely that `SystemQuery` won't be needed anymore, too.

I don't forsee any more major API changes beyond completing this rework, and legion v1.0.0 will likely closely resemble this, after which things should stabilise and breaking changes will hopefully be rare.

TomGillen commented on legion_experiment/src/query/view.rs:
```diff
@@ -0,0 +1,349 @@
+    unsafe fn find_components_mut<T: Component>(&mut self) -> Option<&mut [T]>;
+}
+
+pub trait ReadOnlyFetch<'a>: Fetch<'a> + ReadOnly {
```
`ReadOnly` and `ReadOnlyFetch` are both marker traits which indicate views (and their fetches) which do not perform any mutable accesses. It is used as a constraint in e.g. `Query::iter`, so that you can run a query with only `&World` with safe code, but only when you are not writing to any components.

These traits should probably be marked `unsafe`, actually (much like `Send` and `Sync` for similar reasons).

TomGillen commented on legion_experiment/src/query/view.rs:
```diff
@@ -0,0 +1,349 @@
+    _borrow: Exclusive<'a>,
+}
+
+impl<T: Component> ReadOnly for Write<T> {}
```
Yep, that is a pretty major mistake.

Guvante commented on legion_experiment/src/storage/filter.rs:
```diff
@@ -0,0 +1,596 @@
+                #![allow(non_snake_case)]
+                let ($( $ty, )*) = self.filters;
+                And {
+                    filters: ($( $ty, )* rhs),
```
```suggestion
                    filters: (self, rhs),
```

Guvante commented on legion_experiment/src/storage/filter.rs:
```diff
@@ -0,0 +1,596 @@
+            #[inline]
+            fn bitor(self, rhs: Rhs) -> Self::Output {
+                Or {
+                    filters: (self, rhs),
```
```suggestion
                    filters: ($( $ty, )* rhs),
```

ooflorent commented:
It looks like there are some confusions around `matches_layout`: sometimes it is called with _components_ first, sometimes with _tags_.

TomGillen commented:
I have just pushed something I have been messing about with for some time. 

## Packed Archetypes
The largest change is the introduction of what I am calling "packed archetypes". Legion can now optimise the layout of its internal storage for specific access orders as defined by pre-defined component groupings. Component slices for each archetype _may_ be packed in-order into a single memory allocation if the world determines it to be worthwhile. Packed archetype slices can shrink and grow in-place but, if they run into the next slice, they will be "unpacked" into their own vec. This is very similar to having to resize a vec as items are added. An archetype slice will only be considered a candidate for packing if the slice has remained stable (i.e. no entities added or removed) for a certain amount of time. I would expect this to include the vast majority of the entities in most typical game scenes. This decision is made per component type, not per archetype. Each component storage maintains a "fragmentation" heuristic, which is an estimate of the number of cache misses which would be avoided if the storage were re-packed, and this is used to determine if a component should be packed.

Component groups are specified via a tupl;e of component types (e.g. `<(A, B, C)>::into_group()`, which will accelerate queries which _exactly_ ask for `(A)`, `(A, B)`, or `(A, B, C)`. Queries which do not match a group are not negatively impacted relative to how legion has always performed without this feature.

The largest code change as a result of this feature is that all components of a specific type are now owned by a single component storage per world, rather than each archetype holding onto its own component storage. Archetypes are represented as slices of components whereby each index across all slices refer to the same entity, as they do now, but these slices are no longer divided into fixed size chunks.

## Tags
I have not ported tags over to this architecture. It could be done, but I am not yet decided whether or not to drop tags as a feature. Legion#s internal design is very heavily inspired by Unity's ECS, and I initially found the example use cases Unity gave for their "shared components" to be quite compelling (and, indeed, those examples are still compelling). However, in using both legion and the Unity ECS I have found myself and my colleages _very_ rarely using the feature. More importantly, one of the most commonly asked questions I get from users of legion is "should I use a component or a tag?", and my answer so have has, every time, been "use a component".

## Performance
I have not done a lot of benchmarking of this branch so far, but here are the preliminary results:

```
legion-master           time:   [6.0800 us 6.0970 us 6.1150 us]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

legion-packed-archetypes ordered
                        time:   [2.1003 us 2.1337 us 2.1897 us]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

legion-packed-archetypes unordered
                        time:   [2.0995 us 2.1083 us 2.1168 us]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
```

This benchmark creates entities with 4 different component layouts, `(A,)`, `(A, B)`, `(A, B, C)` and `(A, B, C, D)`, with 1000 entities for each layout. It measures the iteration times for a query over `(&A, &mut B)`.

The experimental branch is **3 times as fast**.

However, there are a few things to note. The first is that the "ordered" benchmark (which define an `(A, B, C, D)` group) is slightly slower than the "unordered" benchmark. The difference is small, but it is consistent. This makes no sense. I would not expect to see much gain from archetype packing when there are only three archetypes, but even so the "ordered" query path is identical except for performing one fewer indirection into a lookup table and its accesses are more linear and ordered.

The second issues is that, upon further investigation, I do not believe that the performance difference between the two branches is actually caused by any of the major changes in this branch....

At its core, a legion query produces an iterator of chunks. Those chunks can then be iterated through to pull out each entity's components. The most "manual" way of doing this is via a nested loop:

```
for chunk in query.iter_chunks(&world) {
   for components in chunk {
      ....
   }
}
```

On the experimental branch, this takes approximately 150ns in one of my benchmarks.

However, users usually don't want to interact with chunks directly, and so we provide wrapper functions which will flatten this nested loop. On the experimental branch, this is simply implemented by wrapping `iter_chunks` inside a `std::iter::Flatten`.  Master uses a custom iterator, but it functions in much the same way. `Flatten` will iterate though the outer iterator, call `IntoIterator` on each value, and then iterate through the inner iterator. It does just what our manual nested loop does.

`query.iter(&world)` takes 800ns.

It would appear that `std::iter::Flatten` is _really slow_. The experimental branch's `Query::for_each` is implemented internally via a nested loop, and I believe this accounts for most of the performance improvement.

## Misc Changes

### Universe Sharding
Universes can be sharded via `Universe::sharded(n: u64, of: u64) -> Universe`. Sharding a universe will divide the Entity ID space between all shards, ensuring that each universe will allocate unique IDs. This can be used to, for example, ensure that entity IDs for serialised content and runtime created entities will never conflict, or to ensure that simulations running in different processes (or different machines) will have unique IDs.

### Entity IDs are no longer generational
We no longer ever use an entity ID as an index, which means we don't need to try and limit the size of the address space via generations. Entity IDs are now defined as `Entity(u64)` and are allocated by incrementing an atomic counter. This greatly simplifies entity ID management, which was a large portion of the entity insertion time on the master branch (about 50% of entity insertion cost).

### Query Caching
Caching of query archetype layout matches is now mandatory and always happens.

### Chunk API
The chunks returned from `iter_chunk` (of which there is now a 1:1 match with archetypes) have a few new APIs:
 * `into_iter()`. Chunks now implement `IntoIterator` which produces an iterator over the view components.
 * `into_components()`. This returns a friendly raw representation of the chunk's internal data. The exact type is determined by the view, but for example a `(Read<A>, Write<B>, TryWrite<C>)` view would return `(&[A], &mut [B], Option<&mut [C]>)`.
 * `archetype(&self) -> &Archetype`. Allows you to retrieve the archetype associated with the chunk. This archetype allows the user to ask about the layout of the entity such as what component types are attached to the entities in the chunk.

### `Entity` is now a `View`.
The `iQuery::ter_entities`, `Query::for_each_entities` and related methods are gone. Instead, users can include `Entity` in their view. e.g. `<(Entity, Read<A>, Write<B>)>::into_query()`.

Nehliin commented:
@TomGillen I have been using "model handle" which refers to a unique asset as a tag in my code. The handle itself contains an id which is used to fetch the actual asset. This means I can do chunk iteration of transform data tagged by a model handle and do instance drawing per chunk which is really nice. Here is a short snippet showing it off:
```rust
let mut offset_map = HashMap::new()
let query =
            <(Read<Transform>, Tagged<ModelHandle>)>::query().filter();
for chunk in query.par_iter_chunks(world) {
     // This is guaranteed to be the same for each chunk
     let model = chunk.tag::<ModelHandle>().unwrap();
     let offset = *offset_map.get(model).unwrap_or(&0);
     let transforms = chunk.components::<Transform>().unwrap();
     offset_map.insert(model.clone(), offset + transforms.len());
     let model = asset_manager.get_model(model).unwrap();
     render_pass
          .draw_model_instanced(model, offset as u32..(offset + transforms.len()) as u32);
}
```
How could I achieve something similar using components? If I would use a unique model handle type per asset I would have to have one query per unique model which is certainly not maintainable in the long run.

kabergstrom commented:
Tags are like custom primary keys, which is very useful for modelling data layout. Recommended usage is probably approximately similar to primary keys in RDMSs, except point lookups are obviously not fast.

cart commented:
I'm building a general purpose game engine with legion and I do not use tags anywhere. I agree that being able to iterate over "chunks" having the same asset id tags seems powerful on paper, but in practice i don't _just_ care about iterating over sets of entities that use the same model/texture pair. I also care about entity position in the world so I can draw back-to-front or front-to-back. This means that I'm accessing entities randomly anyway.

I also consider tags to be confusing for users. I abstract tags out entirely in my "WorldBuilder" api because i believe they negatively affect legibility and introduce cognitive load . I'm sure theres a use case for tags, I just haven't personally encountered it. Ideally tags would be optional / only a part of the api when you need them.

TomGillen commented:
> I don't really like the idea of the storage system doing things like repacking automatically based on heuristics. But my understanding is that the grouping feature is opt-in?

Yes. Repacking only happens if you call `World::pack`, and you can configure how it decides when a component storage is sufficiently fragmented to be repacked, and how each archetype within is considered eligable for packing.

Guvante commented on legion_experiment/src/storage/packed.rs:
```diff
@@ -0,0 +1,603 @@
+    fn new(min_capacity: usize) -> Self {
+        if size_of::<T>() == 0 {
+            Self {
+                ptr: NonNull::new(std::mem::align_of::<T>() as *mut _).unwrap(),
```
I think these can be changed to `NonNull::dangling()`

cart commented:
It would be great to be able to use Entity(u64) as a stable id for serialization / networking. Currently that isn't possible because entity id allocation is fully abstracted out from the user. It seems like this would be possible with the following changes:
* EntityAllocator becomes a trait. By default it could still be an ascending atomic counter, but there could be an alternative impl that avoids collisions.
* Expose an api that lets users instantiate entities with explicit Entity values. Maybe something like `world.extend_with_ids(vec![(A,B), (A,B)], vec![Entity(0), Entity(1)])`

Of course users could always define their own EntityId component, but that seems a bit redundant given that the definition of Entity is "unique identifier of entity in world" and the definition of EntityId would also be "unique identifier of entity in world". It would also mean that I would need to build a higher level wrapper over entity insertion to ensure that all entities have the EntityId component.

cart commented:
Last message on the subject because I dont want to spam you all with potentially off topic information. I just created a "GuidEntityAllocator" in my legion fork that generates a new random id for each entity and allows "queuing up" new Entity ids. Nothing appears to be broken in my relatively complex project and I was able to write a serde implementation that preserves entity ids.

It works well enough for my use cases that I will probably roll with it regardless of the choices made here, but it would be nice to support this in upstream legion.

TomGillen commented:
These changes are ready to merge with master. As this PR essentially amounts to a rewrite of the library, I do not want to do this unilaterally and want feedback and more eyes on the changes first.

To anyone who gives this a look, I would recommend you checkout the code and run `cargo doc --no-deps --show`. The new documentation provides a much better introduction than it used to.

@kabergstrom and @aclysma The internal storage model changed a bit (due to archetype packing and the removal of tags), and both world merging and serialization have also changed as a result. This will produce large breaking changes on the prefab stuff you have been working on. However, most of the breakage is in the form of simplifications. What I really want your feedback on are the changes to legion's merging and serialization. Both of these now do more out of the box, and I need to know if they still offer what you need.

To summarize some of those changes:
 * World merging now handles entity ID conflicts itself, and is configurable between a few policy options on how to handle conflicts. There is more subtlety to doing this correctly than just checking to see if an entity with the same ID already exists in the destination world, and the new code should handle those cases correctly.
 * World merging can, with the two provided `Merger` implementations, both `move` (via the `Move` merger) components and `clone` them (via the `Duplicate` merger). The `Duplicate` merger is also capable of performing component type conversions.
 * `merge_from` accepts a map of entity IDs for manual mapping (for parity with master), but I don't know if you will still need it with other changes. I would rather remove this if it is not needed.
 * Universe sharding might have implications on how prefabs and serialized scenes should work.
 * Serialization has a default "world serializer" impl in the `Registry`. This type provides a fair bit of functionality and you might be able to just use that and not need to implement your own.
 * Deserialization cannot deserialize into an existing world (it always creates a new one) and it performs no entity ID remapping. Remapping should be handled by the world merge, which is the point at which conflicts can actually happen.
 * There are two serialization layouts. This is largely transparent in that the serializer selects an appropriate layout based upon whether or not the output format is supposed to be human readble, and the deserializer can read both layouts.

There are examples of what each of these looks like in the docs for `World::as_serializable` and `World::merge_from`.

Human readbale serialization output looks something like this in JSON (using a string as the Registry's type ID):
```json
{
  "_meta": {
    "component_groups": [],
    "entity_id_next": 128, 
    "entity_id_offset": 0, 
    "entity_id_stride": 1  
  },
  "entities": {
    "124": {
      "isize": 8,
      "usize": 8
    },
    "125": {
      "isize": 7,
      "usize": 7
    },
    "126": {
      "isize": 6,
      "usize": 6
    },
    "127": {        
      "isize": 5,   
      "usize": 5    
    },
    "60": {
      "bool": false,
      "isize": 4,   
      "usize": 4    
    },
    "61": {
      "bool": false,
      "isize": 3,   
      "usize": 3
    },
    "62": {
      "bool": false,
      "isize": 2,
      "usize": 2
    },
    "63": {
      "bool": false,
      "isize": 1,
      "usize": 1
    }
  }
}
```

This puts components next to each entity. It is intended to be easy to read and manually edit, and to work well with VCS (e.g. to keep changes local in the diff).

If you serialize with something like `bincode`, however, it will write with a different layout closer to what it currently does on master - it will write all components of each type together. This is significantly more efficient to deserialize.

kabergstrom commented:
Cool stuff Tom! There are lot of changes and it'll take time to look through it, but based on your comments I'd like to raise an issue:
> Deserialization cannot deserialize into an existing world (it always creates a new one) and it performs no entity ID remapping. Remapping should be handled by the world merge, which is the point at which conflicts can actually happen.

This would require world allocation for every deserialize operation, right? The method I implemented always allocates new entity IDs when deserializing, and keeps a map of <SerializedID, Entity> - why did you change it?

kabergstrom commented:
Another issue I noted:

You've implemented Clone and Serialize for Entity, but it does not handle remapping IDs, and as per above does not allow for custom serialized format. So you still can't use Entity for references to other entities, since they will become invalid after a clone/move with ID conflict.

TomGillen commented:
Serialization and deserialization preserve entity IDs through a round trip, and the IDs are valid in the deserialized world (and won't cause conflicts with later native allocations in that world).

The only place where it is possible to have an ID conflict is when merging two worlds. The new merge implementation does this much more robustly than the old, and it does both support providing a manual ID mapping (which I am not sure you actually need any more) and it returns a map which communicates which IDs have changed and what they were changed to.

With universe sharding, it may also be possible to separate serialized content, remote entities (owned by a network client) and local runtime created entities such that they will never conflict, which would obviate the need to remap in the first place.

The idea was that you would deserialize and then merge. A way to shortcut that chain and avoid the intermediate allocation might be worthwhile, though.

TomGillen commented:
I should also mention that the component storage types are now strongly typed (although you can still interact with them with code that doesn't know the type). FFI support is still planned (including the storage of external types), but there are a few possible levels of support and a few ways we could do things, so it needs some design work. As legion didn't expose a finished FFI interface, I don't consider this to be a regression.

kabergstrom commented:
> With universe sharding, it may also be possible to separate serialized content, remote entities (owned by a network client) and local runtime created entities such that they will never conflict, which would obviate the need to remap in the first place.

Could you elaborate on this?

My primary concern is that it won't be possible to have IDs that are stable across deserialize-merge-serialize cycles. This happens commonly when loading an entity from disk, placing it in a world and then serializing it back again.

Without a stable ID, persisted references can become invalid. Another concern I have is that, with no way to customize the Entity serialization, it won't be possible to have more complex persisted references. What I mean with this is a reference to an entity in a persisted file for example - it will be impossible to know what a single u64 means in the global namespace of a project.

TomGillen commented:
If you create an entity, serialize the world, and then deserialize it, the entity ID will be the same. IDS are stable through round trips through serialization.

Merging is the only point where a conflict can occur, and the merge operation detects and handles that. It does allow you to manually provide remappings (but I am not sure you need that any more) and it also returns a map of IDs that have been changed so you can track those changes.

Universe sharding lets you divide the ID address space between universes. So you can create a universe as "1 of 2" offline, and universe "2 of 2" at runtime, and entities loaded from the offline universe are guarenteed to not conflict with entities allocated in the runtime universe.

See the docs for `World::as_serializable`, `World::merge_from`, `EntityPolicy`, and `Universe`.

kabergstrom commented:
> Merging is the only point where a conflict can occur, and the merge operation detects and handles that. It does allow you to manually provide remappings (but I am not sure you need that any more) and it also returns a map of IDs that have been changed so you can track those changes.

This is what I mean - if a remapping event occurs, how will I ensure my Entity references are valid in the new world?

kabergstrom commented:
> If you create an entity, serialize the world, and then deserialize it, the entity ID will be the same. IDS are stable through round trips through serialization.

The entity ID will be the same, but the serialized ID can conflict with entities serialized in other files that were serialized during another session/in another world. This makes each serialized world have its own local key space for entity IDs, meaning more information is needed for cross-file references. It seems like the API leaves it up to the user to keep extra metadata about these references, but it's not possible to customize the serialization or cloning of references to patch these up as necessary

TimonPost commented:
Skimmed through the changes, I had some questions/remarks.

> Human readable serialization output looks something like this in JSON

When I want to send the world state over the network the human-readable aspect doesn't really matter to me. Would there be a way for a legion to support both human-readable and non-readable? Or would you prefer to keep this functionality up to the user?

> Registry

The name of the 'Registry' sounds a bit confusing to me. I would not directly associate it with a world serializer. Maybe another name would be better here. 

> crossbeam-events feature flag

Why would you enable this one by default? Is this a feature that many people are going to be using? Secondly, I am not sure how this is handled currently, but if I don't read the messages from the channel will this be an evergrowing buffer?

> Serialize

Why would you want this to be enabled by default? I am not sure if this would be a main usecase.

TomGillen commented:
> When I want to send the world state over the network the human-readable aspect doesn't really matter to me. Would there be a way for a legion to support both human-readable and non-readable? Or would you prefer to keep this functionality up to the user?

If you serialize with something like bincode, it will use a more compact layout which is more efficient to deserialize.

> The name of the 'Registry' sounds a bit confusing to me. I would not directly associate it with a world serializer. Maybe another name would be better here.

Any suggestions? It was named `Registry` because it is where your register all of your types.

> Why would you enable this one by default? Is this a feature that many people are going to be using? Secondly, I am not sure how this is handled currently, but if I don't read the messages from the channel will this be an evergrowing buffer?

It is enabled by default because it is enabled by default in master (you cant opt-out). Currently, master requires a `crossbeam_channel::Sender` in its public event API, whereas the new code has moved that under a trait. Whether or not the channel would buffer indefinitely without being consumed would depend on what the `EventSender` impl does. For a crossbeam `unbounded` it would buffer events indefinitely, whereas for a crossbeam `bounded` it will drop new events when the buffer fills.

> Why would you want this to be enabled by default? I am not sure if this would be a main usecase.

Outside of libraries and tech demos, I am not sure there would be many cases where you _wouldn't_ need to persist scene data and/or load it from disk. Even so, making more parts of the library opt-in might be worth considering. It will be easier to make features default in the future than to do the reverse.

aclysma commented:
Some comments so far (just on serialization, haven't tried fixing clone merge yet)
 * No longer supports serializing other types of IDs. We now must serialize a uuid/entity lookup alongside the world. Before, the UUIDs were inline. 
 * When deserializing, a temporary of all components is created and then copied into legion storage. Before, deserialization was in-place and did not need this copy. This is also creating unnecessary heap allocations (the temporary Vecs).
 * The registry system is not useful for us as we will still need a registry for other non-legion functionality and would prefer not to have the unnecessary additional one. 
 * The “Wrapper” type is required to implement deserialization via our own registry but is not public. (I think WorldSeed is a better name)
 * The old trait was more flexible and less code. In my opinion it was better in every way. 
 * The new registry code may be helpful to some people and I don’t see a problem with it being the “canonical” way to serialize legion worlds. However, legion core should expose lean, minimal hooks to implement this kind of functionality. (Such as the trait that was removed.) This way, the inevitably opinionated serialization implementations can have thin, efficient access to do what they need to do and people who struggle with that API will have an easier time using it. I personally would suggest putting it in an optional downstream crate, built on top of a lean interface, such as the trait that was removed. 

At the risk of going off topic I also want to make some comments on using legion serialization for network synchronization of data because it's been brought up a couple times in passing by multiple people. For absolutely trivial games, this might be fine. (Like something turn-based where it's ok to sync the entire world state to everyone all the time on a TCP socket.) But anything non-trivial will require far more control and completely different data representations than what you would want to put in a legion world.

aclysma commented:
Also, I'm quite sad to see tags go. I think the main source of friction was that many APIs required tags as an argument when most people won't need them. If there was a non-tags and tags version of a few key functions (the non-tags could trivially call into the tags one passing ()), I don't think it would have been distracting in the way that they were. If it's possible to get the same functionality as tags by building on top of a core without them (i.e. spawn a world per tag combination) without overhead then that might be a better design (in keeping with the philosophy that core legion should do as little as possible.) But I'm not sure if the same functionality can be achieved without also adding overhead that wasn't there in the original implementation.

aclysma commented:
Unfortunately I couldn't get very far with clone merge, because now it takes the source world as a mutable parameter. I could Arc<Mutex<T>> my source so that I can get a mutable reference to it and placate the compiler, but there's really nothing stopping the source world from being changed, which is explicitly not what we want here. Possibly I could also clone the source world into a new temporary, or deserialize from a byte stream a new instance of a world, but these are pretty wasteful.

TomGillen commented:
What are you trying to do with clone merge?

aclysma commented:
I want to load a legion world from disk and hold it in an immutable state. It would not be part of the simulation.
 * If we want to spawn an instance of the prefab, we can do so by clone merging from that immutable world into the serialization world.
 * We might also clone merge that prefab into a temporary copy, modify it, and then use serde_diff to detect changes that were made and wrap them into a transaction. This drives our undo/redo support. The diffing approach has other advantages described in the link below. 

In the first case, we might convert editor-representations of a component into a runtime representation. In the second case, it's a pure copy. I don't think we have any case where we would want the source data to be modified.

More details here: https://community.amethyst.rs/t/atelier-legion-integration-demo/1352
The crate that implements this: https://github.com/kabergstrom/prefab

TomGillen commented:
And does the `Duplicate` merger not allow for this?

aclysma commented:
Functionally it might, but clone_from and merge_from both take the source world by mutable reference. Probably because it's sharing the codepath with Move. Merger::merge_archetype passes the source world along to those impls via mutable reference.

TomGillen commented:
Yea, it might make more sense to separate the two further to allow duplicate to only require `&World`. Does the functionality do what you need it to though? That type was designed to allow what I thought you were doing with the prefab stuff, but without requiring the user to implement so much code themselves before they could use it, but I might have overlooked something.

aclysma commented:
I think it will but it's hard to say without trying it. It did feel like there was a case for both a Merger and MergerMut but it might increase the code required by quite a bit. For what it's worth, I'm not planning to use either the Move or Duplicate impls of Merger. Similar to serialization we already have a registry for other reasons (like hooks for serde_diff and atelier assets). The prefab crate provides its own interfaces for downstream code to specify how to clone/transform data. The Merger trait looks like it could do the job but I'd like to implement it to make certain. :)

TomGillen commented:
@aclysma and @kabergstrom 

I have reworked how entity IDs, serialization and merging work in response to your feedback. It all makes more sense when you see how it all fits together, but I'll try and explain:

**Entitites**:  
* `Entity` IDs are opaque structures. They identify a location in a world at which an entity may exist. Users cannot create their own `Entity` structs.
* Any given world may or may not contain a specific `Entity`, and that may change as the world is mutated.
* `Entity` IDs are meaningless outside of the process.

**Canon Names**:  
* Entities can be assigned a canonical name (a 16 byte UUID).
* This is the only form of "strong identity" that an entity can have.
* A given name will always be associated with the same given `Entity` ID within the same universe.
* An entity pushed into two worlds with the same name will be assigned the same `Entity` in both worlds.
* "Canonized" entities are the only way that two entities in different worlds can be assigned the same identity.
* Any attempt to push an entity with a name into a world which already has that name (via `push_named`, `universe.canon_mut().canonize`, merging, or deserialization) will cause the entity in the destination to be _replaced_ with that defined in the source. As names as the only way two entities can have the same identity, this is the only way two entities can "conflict".

**Serialization**:  
* As `Entity` is meaningless outside of the process, serialized entities are always identified via their canon name.
* Entities without a name will be assigned one during serialization.
* These names round trip correctly. i.e. If you create a world, serialize it, deserialize it in another process, modify the world, and serialize it again, the entity names will be stable.

**`Entity` Serialization**:  
* The `Entity` struct, wherever it appears inside a component as a reference to another entity, will serialize itself as the target entity's name.
* The target entity will be canonized if it does not yet have a name.
* Upon deserialization, the name will be resolved to an `Entity` ID.
* The above works, **even if the target entity is not defined in the file, or does not yet exist in any runtime world**.
* When an entity with the same name as that referenced is at any point created, it will be created with the same `Entity` ID as was deserialized, and no other entity will ever be created with that ID.

I have also made a few miscellaneous changes: fixed a few types not being exported, separated `clone_from` and `copy_from` code paths, `clone_from` no longer requires `&mut World`, etc.

TomGillen commented:
JSON (human readable) serialization format now looks like:

```json
{
  "entities": {
    "14ab8211-4a7c-4817-813b-8acf282c27de": {
      "bool": false,
      "isize": 4,
      "usize": 4
    },
    "614a1c19-3718-4d30-b249-04d5c45df547": {
      "entity_ref": "d6be1905-3714-40be-8226-60f3177ca395",
      "isize": 6,
      "usize": 6
    },
    "9d51f554-92c5-446b-b764-5892be5dd14a": {
      "bool": false,
      "isize": 3,
      "usize": 3
    },
    "c8ca7a5c-834b-4f7e-a359-68a3a3779ed7": {
      "entity_ref": "d6be1905-3714-40be-8226-60f3177ca395",
      "isize": 7,
      "usize": 7
    },
    "d3ef5175-d430-4f15-ab16-9b6082bfeefd": {
      "bool": false,
      "isize": 2,
      "usize": 2
    },
    "d6be1905-3714-40be-8226-60f3177ca395": {
      "bool": false,
      "isize": 1,
      "usize": 1
    },
    "ea27971a-9c71-41a6-babb-4c1a47418c77": {
      "entity_ref": "d6be1905-3714-40be-8226-60f3177ca395",
      "isize": 8,
      "usize": 8
    },
    "fa5f9c08-a308-4555-a7e5-11121d2e499f": {
      "entity_ref": "d6be1905-3714-40be-8226-60f3177ca395",
      "isize": 5,
      "usize": 5
    }
  }
}
```

Where "isize", "usize" "bool" and "entity_ref" are the external names registered with the `Registry` for the component type IDs (these would probably be UUIDs in a proper application).

TimonPost commented:
Json serializer/deserializer seems to work good, however, bincode deserializer doesnt seem to deserialize very well.

```

    let serialized = serde_json::to_value(&world.as_serializable(any(), &registry)).unwrap();
    println!("{:#}", serialized.to_string());

    let universe = Universe::new();
    let world: World = registry
        .as_deserialize(&universe)
        .deserialize( &mut serde_json::Deserializer::from_str(&serialized.to_string())).unwrap();

    println!("Number of entities in deserialized world {}\n\n", world.len());

    let serialized = bincode::serialize(&world.as_serializable(any(), &registry)).unwrap();
    println!("{:?}", serialized);

    let universe = Universe::new();
    let world: World = registry
        .as_deserialize(&universe)
        .deserialize( &mut bincode::Deserializer::from_slice(&serialized, bincode::DefaultOptions::default())).unwrap();

    println!("Number of entities in deserialized world {}", world.len())
```

Any ideas why this is?

TomGillen commented:
@TimonPost Your bincode example won't deserialize correctly because you serialized with `bincode::serialize` and deserialized using deserialize with `bincode::DefaultOptions::default()`. Believe it or not, but bincode's default serialization options are _not_ `DefaultOptions::default()`. You need to also used fixed int length encoding and allow trailing bytes. Not setting those configuration options causes bincode to just read nonsense and often without erroring (in this case, some of the sequences are being read as empty).

You need to use:

```rust
bincode::DefaultOptions::new()
    .with_fixint_encoding()
    .allow_trailing_bytes()
```

TimonPost commented:
I again came across weird behavior in my bincode serializing code. When a type id is not known it will use `IgnoreAny` [here](https://github.com/TomGillen/legion/pull/115/files#diff-613f8cc135f1762b6483319d71983843R406). `IgnoreAny` is not supported by bincode and will throw an error (`Bincode rust does not support Serializer::deserialize ignored any`).

This is done in the following two places:

https://github.com/TomGillen/legion/pull/115/files#diff-fb7fb3d74cc05a5c59858d2d6b136bd8R249
https://github.com/TomGillen/legion/pull/115/files#diff-613f8cc135f1762b6483319d71983843R406

While the legion logic is correct logic there is some underlying issue considering the nature of `TypeId` and its unique hash when using legion across multiple binaries. 

I use two separate binaries who use the same registered types and same logic for constructing the registry with the types but somehow the `SerializedTypeId`s are different for each binary. Those two binaries are two endpoints communication over the interwebs. Upon receipt of the world state, it should deserialize it and apply it to the local game world. Because of the deferring `SerializeTypeId`'s this code will error out with bincode. This is not only a problem with bincode but with any other serializer, it will ignore the deserializing and thus I end up with an empty entity in my world. 

I am not sure if this is something to care about. But if legion is used across multiple binaries/network using `TypeId` can be misleading. If this is know I would suggest placing a comment it can save a lot of confusing for users.

kabergstrom commented:
@TimonPost SerializableTypeIds have this comment:

/// A default serializable type ID. This ID is not stable between compiles,
/// so serialized words serialized by a different binary may not deserialize
/// correctly.

I think it's a bad default to include in the project though, since it's more or less useless for what you'd expect it to be.

As for `IgnoreAny`, bincode is not self-delimiting so it's impossible to deserialize something without knowing the exact type layout used when serializing. Unfortunately this is a limitation of the format, if you want something more robust, I'd suggest messagepack.

kabergstrom commented:
I recommend using https://docs.rs/type-uuid/0.1.2/type_uuid/ for your serialized type ID

TomGillen commented:
I am going to merge this PR. Further feedback should be better served with more focused and specific issues and PRs. I am going to give it a bit more time for us to discover and work through issues with this rewrite before publishing 0.3 to crates.io.
